File: Tools.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.*;


public class Tools {
    public static int top(int[] arr)
    {
        if (arr.length==0) {
            return 0;
        }
        if(arr[arr.length-1]==-1)
            return -1;

        if (arr.length==1&&arr[0]==-1) {
            return -1;
        }
        if (arr.length==1&&arr[0]!=-1) {
            return 0;
        }
        boolean flag= false;
        if(arr[0]==-1)
            flag = true;
        for (int i = 0; i < arr.length-1; i++) {
            if(arr[i+1]!=-1)
               { 
                    if(flag)
                        return i+1;
                    return i;
               }
        }
        return -1;
    }
    public static boolean full(int []arr)
    {
        return arr[0]!=-1;
    }
    public static boolean isValidMove(int i1,int i2,int[][] c1) {
        if (top(c1[i1])==-1) {
            return false;
        }
        return top(c1[i2])==-1||(c1[i1][top(c1[i1])]==c1[i2][top(c1[i2])]&&!full(c1[i2]));
    }
    public static void performMove(int fromContainer, int toContainer, Container container) {
        int[][] levels = container.getLevels();
        int color = levels[fromContainer][top(levels[fromContainer])];
        while (top(levels[fromContainer])!=-1&&levels[fromContainer][top(levels[fromContainer])]==color&&!full(levels[toContainer])) {
            levels[fromContainer][top(levels[fromContainer])] =-1;
            if (top(levels[toContainer])==-1) {
                levels[toContainer][levels[toContainer].length-1] = color;
            }
            else{
                levels[toContainer][top(levels[toContainer])-1] = color;
            }
        }
        container.setLevels(levels);
    }
    public static boolean isGoalState(Container c) {
        int cur;
        for (int i = 0; i < c.getLevels().length; i++) {
            cur = c.getLevels()[i][0];
            for (int j = 0; j < c.getLevels()[0].length; j++)
                if (cur!=c.getLevels()[i][j]) {
                    return false;
                }
        }
        return true;
    }
    public static int calculateHeuristic(Container c) {
        int heuristicValue = 0;
        int[] res = new int[c.getNumOfColors()];
        int color,top;
        int [][] levels = c.getLevels();
        for (int i = 0; i < levels.length; i++) {
            top = top(levels[i]);

            if (top!=-1) {
                heuristicValue+=2;
                for (int j = levels[i].length-1; j > top; j--) {
                    color = levels[i][j];
                    while (levels[i][j-1]==color) {
                        j--;
                        if(j==0)
                        break;
                    }
                    res[color] += j - top;
                }
            }
        }
        for (int i = 0; i < res.length; i++) {
            heuristicValue += res[i];
        }
        return heuristicValue - res.length*2;
    }
    public static void print2DArray(int[][] array) {
        for (int i = 0; i < array.length; i++) {
            for (int j = 0; j < array[i].length; j++) {
                System.out.print(array[i][j] + " ");
            }
            System.out.println();
        }
    }
    public static void printArray(int[] array) {
        for (int i = 0; i < array.length; i++) {
            System.out.print(array[i] + " ");
        }
        System.out.println();
    }
    public static int[][] filterArrays(int[][] arr) {
        ArrayList<int[]> filteredList = new ArrayList<>();
        boolean hasAddedFullOfNegOnes = false;

        for (int[] subArr : arr) {
            boolean isFullOfNegOnes = true;
            for (int num : subArr) {
                if (num != -1) {
                    isFullOfNegOnes = false;
                    break;
                }
            }
            if (isFullOfNegOnes && !hasAddedFullOfNegOnes) {
                filteredList.add(subArr);
                hasAddedFullOfNegOnes = true;
            } else if (!isFullOfNegOnes) {
                filteredList.add(subArr);
            }
        }
        int[][] result = new int[filteredList.size()][];
        for (int i = 0; i < filteredList.size(); i++) {
            result[i] = filteredList.get(i);
        }
        return result;
    }
    public static int[][] addNewArrays(int[][] arr, int newArrays) {
        int numRows = arr.length;
        int numCols = arr[0].length;
        int[][] result = new int[numRows + newArrays][numCols];
        for (int i = 0; i < numRows; i++) {
            result[i] = Arrays.copyOf(arr[i], numCols);
        }
        for (int i = numRows; i < numRows + newArrays; i++) {
            Arrays.fill(result[i], -1);
        }
        return result;
    }
    public static List<Container> generateNeighbors(Container currentContainer) {
        List<Container> neighbors = new ArrayList<>();
        int numContainersStart = currentContainer.getLevels().length;
        
        currentContainer = new Container(filterArrays(currentContainer.getLevels()),currentContainer.getSteps(),currentContainer.getHeuristic(),currentContainer.getNumOfColors());

        int numContainers = currentContainer.getLevels().length;
        for (int from = 0; from < numContainers; from++) {
            for (int to = 0; to < numContainers; to++) {
                if (from!=to&&isValidMove(from, to, currentContainer.getLevels())) {
                    Container neighbor = new Container(currentContainer);
                    performMove(from, to, neighbor);
                    neighbor.setLevels(addNewArrays(neighbor.getLevels(), numContainersStart-numContainers));
                    neighbor.setSteps(currentContainer.getSteps()+1);
                    neighbor.setHeuristic(Tools.calculateHeuristic(neighbor));
                    neighbors.add(neighbor);
                }
            }
        }
        return neighbors;
    }
    public static void outputSolutionPath(Container goalContainer, Map<Container, Container> parentMap, long elapsedTime,int i) {
        System.out.println("Puzzle has been solved in " + goalContainer.getSteps() + " steps in " + elapsedTime/1000000000.0 + " seconds in "+i+" itaretions.");
    }
    public static void updateOpenSet(Hashtable<String, Container> openSet, Set<String> closedSet, List<Container> neighbors) {
        for (Container neighborContainer : neighbors) {
            String neighborStateHash = neighborContainer.toString();
            if (!closedSet.contains(neighborStateHash)&& !openSet.containsKey(neighborStateHash))
                openSet.put(neighborStateHash, neighborContainer);
        }
    }
    public static Container getMinCostContainer(Hashtable<String, Container> openSet) {
        Container minContainer = null;
        double f;
        double minVal = Double.MAX_VALUE;
        for (Container container : openSet.values()) {
            f = container.getHeuristic()+container.getSteps();
            if (f < minVal) {
                minVal = f;
                minContainer = container;
            }
        }
        return minContainer;
    }
    public static int[][] convertStringTo2DArray(String input) {
        String[] arrayStrings = input.split("\\],\\[");
        int maxLength = 0;
        for (String arrayString : arrayStrings) {
            String[] elements = arrayString.replaceAll("[\\[\\]]", "").split(",");
            maxLength = Math.max(maxLength, elements.length);
        }
        int[][] result = new int[arrayStrings.length][maxLength];
        for (int i = 0; i < arrayStrings.length; i++) {
            String[] elements = arrayStrings[i].replaceAll("[\\[\\]]", "").split(",");
            int[] intElements = new int[elements.length];
            for (int j = 0; j < elements.length; j++) {
                if (!elements[j].isEmpty()) {
                    intElements[j] = Integer.parseInt(elements[j]);
                }
                else
                {
                    intElements = new int[0];
                }
            }
            result[i] = fillArrayFromTop(intElements, maxLength);
        }
        return result;
    }

    public static int[] fillArrayFromTop(int[] arr, int maxLength) {
        int[] result = new int[maxLength];
        int len = arr.length;
        int diff = maxLength-len;
        for (int i = len-1; i >= 0; i--) {
            result[i + diff] = arr[i];
        }
        for (int i = 0; i < diff; i++) {
            result[i] = -1;
        }
        return result;
    }
    public static List<String> extractInitRows(String filePath) throws IOException {
        List<String> initRows = new ArrayList<>();
        BufferedReader reader = new BufferedReader(new FileReader(filePath));
        String line;
        StringBuilder section = new StringBuilder();

        while ((line = reader.readLine()) != null) {
            if (line.startsWith("####")) {
                if (section.length() > 0) {
                    initRows.add(extractInitRow(section.toString()));
                    section.setLength(0); // Clear the section
                }
            } else {
                section.append(line).append("\n");
            }
        }
        if (section.length() > 0) {
            initRows.add(extractInitRow(section.toString()));
        }

        reader.close();
        return initRows;
    }
    private static String extractInitRow(String section) {
        String[] lines = section.split("\n");
        for (String line : lines) {
            if (line.trim().startsWith("init=")) {
                return line.trim().split("=")[1];
            }
        }
        return "";
    }
    public static Hashtable<String, Container> sortAndFilterContainers(Hashtable<String, Container> openSet) {
        // Convert the Hashtable values to a List
        List<Map.Entry<String, Container>> entryList = new ArrayList<>(openSet.entrySet());

        // Sort the list based on the heuristic value in descending order
        entryList.sort((e1, e2) -> Integer.compare(e2.getValue().getHeuristic()+e2.getValue().getSteps(), e1.getValue().getHeuristic()+e1.getValue().getSteps()));

        // Calculate the number of containers to keep (1/4 of the total number)
        int numberToKeep = entryList.size() / 4;

        // Filter the containers, keeping the 1/4 with the lowest heuristic values
        List<Map.Entry<String, Container>> filteredEntries = entryList.subList(entryList.size() - numberToKeep, entryList.size());

        // Create a new Hashtable to store the filtered containers
        Hashtable<String, Container> filteredOpenSet = new Hashtable<>();
        for (Map.Entry<String, Container> entry : filteredEntries) {
            filteredOpenSet.put(entry.getKey(), entry.getValue());
        }

        return filteredOpenSet;
    }
    public static void solvePuzzle(Container initialContainer) {
        int i =0;
        List<Container> neighbors =null;
        Hashtable<String, Container> openSet = new Hashtable<>();
        Set<String> closedSet = new HashSet<>();
        Map<Container, Container> parentMap = new HashMap<>();
        openSet.put(initialContainer.toString(), initialContainer);
        long startTime = System.nanoTime();
        while (!openSet.isEmpty()) {
            i++;
            Container currentContainer = Tools.getMinCostContainer(openSet);
            if (Tools.isGoalState(currentContainer)) {
                long endTime = System.nanoTime();
                long elapsedTime = endTime - startTime;
                Tools.outputSolutionPath(currentContainer, parentMap, elapsedTime,i);
                return;
            }
            String stateHash = currentContainer.toString();
            if (!closedSet.contains(stateHash)) {
                closedSet.add(stateHash);
            }
            if (openSet.size()>1000) {
                openSet = sortAndFilterContainers(openSet);
            }
            openSet.remove(currentContainer.toString());
            neighbors = Tools.generateNeighbors(currentContainer);
            Tools.updateOpenSet(openSet, closedSet, neighbors);
        }
        long endTime = System.nanoTime();
        long elapsedTime = endTime - startTime;
        System.out.println("No solution found. Time elapsed: " + elapsedTime/100000000 + " seconds In "+i+" itaretions.");
        return;
    }
}

File: cpp/Tools.cpp
#include <iostream>
#include <vector>
#include <unordered_set>
#include <unordered_map>
#include <queue>
#include <string>
#include <sstream>
#include <fstream>
#include <chrono>
#include <algorithm>

using namespace std;

class Container {
public:
    vector<vector<int>> levels;
    int numOfColors;
    double heuristic;

    Container(vector<vector<int>> levels) : levels(levels), numOfColors(0), heuristic(0) {
        // Determine the number of colors
        unordered_set<int> colors;
        for (const auto& row : levels) {
            for (int color : row) {
                if (color != -1) {
                    colors.insert(color);
                }
            }
        }
        numOfColors = colors.size();
    }

    Container(const Container& other) : levels(other.levels), numOfColors(other.numOfColors), heuristic(other.heuristic) {}

    void setLevels(const vector<vector<int>>& newLevels) {
        levels = newLevels;
    }

    int getNumOfColors() const {
        return numOfColors;
    }

    double getHeuristic() const {
        return calculateHeuristic(*this);
    }

    void setHeuristic(double h) {
        heuristic = calculateHeuristic(*this);
    }

    bool operator==(const Container& other) const {
        return levels == other.levels;
    }

    bool operator!=(const Container& other) const {
        return !(*this == other);
    }

    string toString() const {
        ostringstream oss;
        for (const auto& row : levels) {
            for (int val : row) {
                oss << val << " ";
            }
            oss << "\n";
        }
        return oss.str();
    }
};

// Custom hash function for Container
struct ContainerHasher {
    size_t operator()(const Container& c) const {
        string s = c.toString();
        return hash<string>()(s);
    }
};

int top(const vector<int>& arr) {
    bool flag = false;
    if (arr[0] == -1) flag = true;
    for (size_t i = 0; i < arr.size() - 1; ++i) {
        if (arr[i + 1] != -1) {
            if (flag) return i + 1;
            return i;
        }
    }
    return -1;
}

bool full(const vector<int>& arr) {
    return arr[0] != -1;
}

bool isValidMove(int i1, int i2, const vector<vector<int>>& c1) {
    if (top(c1[i1]) == -1) return false;
    return top(c1[i2]) == -1 || (c1[i1][top(c1[i1])] == c1[i2][top(c1[i2])] && !full(c1[i2]));
}

void performMove(int fromContainer, int toContainer, Container& container) {
    auto& levels = container.levels;
    int color = levels[fromContainer][top(levels[fromContainer])];
    while (top(levels[fromContainer]) != -1 && levels[fromContainer][top(levels[fromContainer])] == color && !full(levels[toContainer])) {
        levels[fromContainer][top(levels[fromContainer])] = -1;
        if (top(levels[toContainer]) == -1) {
            levels[toContainer][levels[toContainer].size() - 1] = color;
        } else {
            levels[toContainer][top(levels[toContainer]) - 1] = color;
        }
    }
    container.setLevels(levels);
}

bool isGoalState(const Container& c) {
    int cur;
    for (const auto& level : c.levels) {
        cur = level[0];
        for (int val : level) {
            if (cur != val) return false;
        }
    }
    return true;
}

int calculateHeuristic(const Container& c) {
    int heuristicValue = 0;
    vector<int> res(c.getNumOfColors(), 0);
    int color;

    for (const auto& level : c.levels) {
        if (top(level) != -1) {
            for (int j = level.size() - 1; j > top(level); --j) {
                color = level[j];
                while (j > top(level) && level[j - 1] == color) {
                    --j;
                }
                res[color] += j - top(level);
            }
        }
    }

    for (int val : res) {
        heuristicValue += val;
    }
    for (size_t i = 1; i < c.levels.size(); ++i) {
        if (top(c.levels[i]) != -1) {
            ++heuristicValue;
        }
    }
    heuristicValue -= res.size() - 2;
    return heuristicValue;
}

void print2DArray(const vector<vector<int>>& array) {
    for (const auto& row : array) {
        for (int val : row) {
            cout << val << " ";
        }
        cout << endl;
    }
}

void printArray(const vector<int>& array) {
    for (int val : array) {
        cout << val << " ";
    }
    cout << endl;
}

vector<Container> generateNeighbors(const Container& currentContainer) {
    vector<Container> neighbors;
    int numContainers = currentContainer.levels.size();

    for (int from = 0; from < numContainers; ++from) {
        for (int to = 0; to < numContainers; ++to) {
            if (isValidMove(from, to, currentContainer.levels) && from != to) {
                Container neighbor(currentContainer);
                performMove(from, to, neighbor);
                neighbor.setHeuristic(calculateHeuristic(neighbor));
                neighbors.push_back(neighbor);
            }
        }
    }

    return neighbors;
}

bool arraysEqual(const vector<int>& arr1, const vector<int>& arr2) {
    return arr1 == arr2;
}

vector<int> findDifferentLevelIndex(const vector<vector<int>>& levels1, const vector<vector<int>>& levels2) {
    vector<int> diffIndex(2, -1);
    bool flag = false;

    for (size_t i = 0; i < levels1.size(); ++i) {
        if (!arraysEqual(levels1[i], levels2[i])) {
            if (!flag) {
                diffIndex[0] = i + 1;
                flag = true;
            } else {
                diffIndex[1] = i + 1;
                break;
            }
        }
    }

    return diffIndex;
}

pair<int, long> outputSolutionPath(const Container& goalContainer, const unordered_map<Container, Container, ContainerHasher>& parentMap, long elapsedTime) {
    pair<int, long> result;
    vector<string> path;
    Container current = goalContainer;

    while (parentMap.find(current) != parentMap.end()) {
        Container parent = parentMap.at(current);
        auto diffIndex = findDifferentLevelIndex(parent.levels, current.levels);
        string action = "Swap container " + to_string(diffIndex[0]) + " with " + to_string(diffIndex[1]);
        path.push_back(action + "\n" + current.toString());
        current = parent;
    }

    reverse(path.begin(), path.end());

    cout << "Puzzle has been solved in " << path.size() << " steps in " << elapsedTime << " milliseconds!" << endl;
    result.first = path.size();
    result.second = elapsedTime;
    return result;
}

void updateOpenSet(unordered_map<int, Container>& openSet, unordered_set<string>& closedSet, const vector<Container>& neighbors) {
    for (const auto& neighborContainer : neighbors) {
        int neighborHashCode = hash<string>()(neighborContainer.toString());
        if (closedSet.find(neighborContainer.toString()) == closedSet.end() && openSet.find(neighborHashCode) == openSet.end()) {
            openSet[neighborHashCode] = neighborContainer;
        }
    }
}

Container getMinCostContainer(unordered_map<int, Container>& openSet) {
    Container minContainer({{}});
    double minHeuristic = numeric_limits<double>::max();

    for (const auto& entry : openSet) {
        double heuristic = entry.second.getHeuristic();
        if (heuristic < minHeuristic) {
            minHeuristic = heuristic;
            minContainer = entry.second;
        }
    }

    return minContainer;
}

vector<vector<int>> convertStringTo2DArray(const string& input) {
    istringstream iss(input);
    string line;
    vector<vector<int>> result;

    while (getline(iss, line)) {
        vector<int> row;
        istringstream rowStream(line);
        int val;
        while (rowStream >> val) {
            row.push_back(val);
        }
        result.push_back(row);
    }

    return result;
}

vector<int> fillArrayFromTop(const vector<int>& arr, int maxLength) {
    vector<int> result(maxLength, -1);
    int len = arr.size();
    int diff = maxLength - len;

    for (int i = len - 1; i >= 0; --i) {
        result[i + diff] = arr[i];
    }

    return result;
}

pair<int, long> solvePuzzle(Container initialContainer) {
    unordered_map<int, Container> openSet;
    unordered_set<string> closedSet;
    unordered_map<Container, Container, ContainerHasher> parentMap;
    openSet[hash<string>()(initialContainer.toString())] = initialContainer;

    auto startTime = chrono::high_resolution_clock::now();

    while (!openSet.empty()) {
        Container currentContainer = getMinCostContainer(openSet);

        if (isGoalState(currentContainer)) {
            auto endTime = chrono::high_resolution_clock::now();
            auto elapsedTime = chrono::duration_cast<chrono::milliseconds>(endTime - startTime).count();
            return outputSolutionPath(currentContainer, parentMap, elapsedTime);
        }

        string stateHash = currentContainer.toString();
        closedSet.insert(stateHash);
        openSet.erase(hash<string>()(stateHash));

        auto neighbors = generateNeighbors(currentContainer);
        updateOpenSet(openSet, closedSet, neighbors);

        for (const auto& neighbor : neighbors) {
            parentMap[neighbor] = currentContainer;
        }
    }

    auto endTime = chrono::high_resolution_clock::now();
    auto elapsedTime = chrono::duration_cast<chrono::milliseconds>(endTime - startTime).count();
    cout << "No solution found. Time elapsed: " << elapsedTime << " milliseconds." << endl;

    return {-1, elapsedTime};
}
bool starts_with(const std::string& str, const std::string& prefix) {
    return str.substr(0, prefix.size()) == prefix;
}
vector<string> extractInitRows(const string& filePath) {
    vector<string> initRows;
    ifstream file(filePath);
    string line;
    stringstream section;

    while (getline(file, line)) {
        if (starts_with(line,"####")) {
            if (section.str().length() > 0) {
                initRows.push_back(extractInitRow(section.str()));
                section.str("");
                section.clear();
            }
        } else {
            section << line << "\n";
        }
    }

    if (section.str().length() > 0) {
        initRows.push_back(extractInitRow(section.str()));
    }

    return initRows;
}

string extractInitRow(const string& section) {
    istringstream iss(section);
    string line;

    while (getline(iss, line)) {
        if (line.find("init=") != string::npos) {
            return line.substr(line.find("=") + 1);
        }
    }

    return "";
}